compiler/passes/convert-typed-uast.cpp

```diff
@@ -0,0 +1,2041 @@
+  void exit(const EmptyStmt* ast, RV& rv) { }
+
+  // traversal cases to do something
+  bool enter(const Module* node, RV& rv);
+  void exit(const Module* node, RV& rv);
+
+  bool enter(const Function* node, RV& rv);
+  void exit(const Function* node, RV& rv);
+
+  bool enter(const Formal* node, RV& rv);
+  void exit(const Formal* node, RV& rv);
+
+  bool enter(const Variable* node, RV& rv);
+  void exit(const Variable* node, RV& rv);
+
+  bool enter(const Literal* node, RV& rv);
+  void exit(const Literal* node, RV& rv);
+
+  bool enter(const Identifier* node, RV& rv);
+  void exit(const Identifier* node, RV& rv);
+
+  bool enter(const Return* node, RV& rv);
+  void exit(const Return* node, RV& rv);
+
+  bool enter(const Call* node, RV& rv);
+  void exit(const Call* node, RV& rv);
+
+  bool enter(const AstNode* node, RV& rv);
+  void exit(const AstNode* node, RV& rv);
+};
+
+TConverter::~TConverter() { }
+
+Expr* TConverter::convertAST(const AstNode* node) {
+  // TODO: is this sufficient?
+  return untypedConverter->convertAST(node);
+}
+
+Expr* TConverter::convertExpr(const AstNode* node, RV& rv) {
+  // traverse to add it to curAList
+  node->traverse(rv);
+
+  // remove the last thing from the current AList and return that
+  return curAList->last()->remove();
+}
+
+ModuleSymbol* TConverter::convertToplevelModule(const Module* mod,
+                                                ModTag modTag) {
+  if (!haveSetupModules) {
+    // create empty modules so we have somewhere to put the functions!
+    setupModulesToConvert();
+    haveSetupModules = true;
+  }
+
+  if (!haveConvertedFunctions) {
+    // convert the functions!
+    convertFunctionsToConvert();
+    haveConvertedFunctions = true;
+  }
+
+  astlocMarker markAstLoc(mod->id());
+
+  ModuleSymbol* modSym = findOrSetupModule(mod->id());
+
+  // TODO: remove this block once the resolver is fully operational
+  if (modSym->modTag != MOD_USER) {
+    return untypedConverter->convertToplevelModule(mod, modTag);
+  }
+
+  topLevelModTag = modTag;
+  convertModuleInit(mod, modSym);
+
+  return modSym;
+}
+
+void TConverter::postConvertApplyFixups() {
+  untypedConverter->postConvertApplyFixups();
+
+  // TODO: apply fixups from this converter
+}
+
+void TConverter::setupModulesToConvert() {
+  // create each module now since we might add a function to it
+  // the order in which this is done is not particularly
+  // important since they will be filled-in in module init order
+  // via calls to convertToplevelModule.
+  for (const ID& modId : modulesToConvertVec) {
+    findOrSetupModule(modId);
+  }
+}
+
+struct SortCalledFnsDeeperFirst {
+  bool operator()(const std::pair<const ResolvedFunction*, int>& a,
+                  const std::pair<const ResolvedFunction*, int>& b) {
+    // reverse sort by depth
+    if (a.second > b.second) return true;
+    if (a.second < b.second) return false;
+
+    // regular sort by ResolvedFunction
+    return *a.first < *b.first;
```

Comment:
IMO supporting `<` is not conceptually complicated. It is just that it impacts many types. Another thought about it is that, it seems somewhat likely to me that we will run into other cases where `<` on some of these types will be needed. One specific case that I can think of: I think we are going to need to sort the instantiations when creating a separate compilation library file (for stable output). But, if we have them sorted, and we have `<` on them, then we can do binary search when looking for a particular instantiation.

Now, I did not implement `<` between types. (I left that as a TODO). How exactly to do that seems somewhere between tricky and annoying to implement, so I am sympathetic to your point.

> can you just sort by ID, and then keep insertion order? Since insertion order is determined by a call graph traversal, I'd expect that to be stable enough

It may be already obvious to you already, but the reason that sorting by ID isn't good enough is that we have to concern ourselves with instantiations of generic functions. It might be possible to track the order in which these were instantiated. We couldn't do that within the types themselves, because that gets into issues with the query framework (on a recompilation, I'd want it to reuse the same instantiation even if it went from being instantiation 1 to instantiation 3).

However it is an interesting idea to include it in the call graph traversal. Were I to do that, would you favor keeping the `<` that I implemented (in case they are useful later) or removing them (so as to avoid any maintenance burden)?

Reply:
I would certainly favor removing them (if you needed them later, you could always find the diff in this PR), but I will not block the PR over their inclusion. If you think they are needed after all (or will be needed in the future), then I trust your judgement.